Normalize images in CategoricalCNNPolicy

maliesa96 opened the pull request:
Normalizes pixel values and accepts uint8 inputs (addresses [#734](https://github.com/rlworkgroup/garage/issues/734))

ahtsan commented on src/garage/tf/misc/tensor_utils.py:
```diff
@@ -63,6 +64,14 @@ def filter_valids(t, valid, name='filter_valids'):
         t, tf.cast(tf.round(valid), tf.int32), 2, name=name)[1]
 
 
+def normalize_pixel_observation(env_spec, obs):
```
I believe we already have this, check out `garage.misc.tensor_utils.normalize_pixel_batch`

maliesa96 replied:
Yep though that takes in a batch of observations and returns a list of normalized observations since it's used by the sampler. This function just accepts one observation. Although `garage.misc.tensor_utils.normalize_pixel_batch` can be rewritten to call `normalize_pixel_observation` in a loop.

ahtsan replied:
I see. Indeed, the difference is probably that one deals with python array while one deals with tf.Tensor.

ahtsan replied:
Would be great if you can provide more information in the docstring. Please refer to [this link](https://github.com/rlworkgroup/garage/blob/master/CONTRIBUTING.md) for guide.

maliesa96 replied:
Thanks, updated. I also followed the docstring format in `garage.misc.tensor_utils.normalize_pixel_batch`.

ahtsan commented:
Could you write a test for the function you added?

maliesa96 commented:
Done.

ryanjulian commented on src/garage/tf/misc/tensor_utils.py:
```diff
@@ -62,6 +63,25 @@ def filter_valids(t, valid, name='filter_valids'):
+    """
+    
+    if isinstance(env_spec.observation_space, gym.spaces.Box):
+        if len(env_spec.observation_space.shape) == 3:
```
what happens if I have a Box(3,) non-pixel observation space?

ryanjulian commented on src/garage/tf/policies/categorical_cnn_policy.py:
```diff
@@ -101,9 +102,9 @@ def __init__(self,
+        state_input = tf.compat.v1.placeholder(tf.uint8,
                                                shape=(None, ) + self.obs_dim)
-
+        state_input = normalize_pixel_observation(self.env_spec, state_input)
```
since the only valid input to this primitive is an image, you should probably verify it's an image here, rather than writing a conditional into the utility function

maliesa96 replied:
Is there a foolproof way of checking whether observations are images, since they are simply tensors? I was going off what was done in `garage.misc.tensor_utils.normalize_pixel_batch` (which also just checks the shape of the Box). I can check that the box boundaries are 255 all around which means they are likely pixels, though that also does not guarantee it. The other option is to a pass an optional flag to the policy that indicates whether or not the observations are images.

ryanjulian replied:
there isn't, and that's one of the difficult parts of this PR.

without adding an `Image` type to akro (which i think is a good idea), there isn't a whole lot which can be done in this PR specifically to guarantee these types are safe.

however, what we *can* do is decide whose job type safety is. since CNN policies can't infer image types, once data reaches them, it must be too late to do anything about it. so it can't be the responsibility of the CNN policies to enforce type safety. similarly, helper functions couldn't possibly do this without types, either (which is the issue I raise here).

if we raise ourselves on level in the call tree, we will find ourselves in the algorithm. that means it must be the algorithm's job to enforce type safety -- it can find out whether the observation is an image by interrogating the environment class, which it actually has access to.

to summarize, there are two good options here. which you use is up to you.

1) Add an Image type to [akro](https://github.com/rlworkgroup/akro) (a package we maintain), get me or @krzentner to cut akro release 0.0.7, bump garage's akro version to 0.0.7, then ensure `GarageEnv` converts the EnvSpec of image-based environments to use the new `Image` type rather than `Box`)
2) Add a flag to the algorithm alerting it that it should expect an image input, check the observation space (inside the algorithm) to ensure it's consistent with an image, refuse to use CNN-based primitives unless this flag is turned on.

(1) Is the long-term solution, but there's something to be said for doing (2) first since it would allow you to close the issue faster.

maliesa96 replied:
Option 1 sounds cleaner to me. I'll create a PR in the akro repo for the new image type.

ryanjulian commented on src/garage/tf/misc/tensor_utils.py:
```diff
@@ -62,6 +63,25 @@ def filter_valids(t, valid, name='filter_valids'):
     return tf.dynamic_partition(
         t, tf.cast(tf.round(valid), tf.int32), 2, name=name)[1]
 
+def normalize_pixel_observation(env_spec, obs):
```
there's a common convention that functions which *might* do something should be called `maybe_foo` rather than just foo.

with that said, see my comments below on why this function shouldn't be checking for input types anyway

ryanjulian commented on tests/garage/tf/misc/test_tensor_utils.py:
```diff
@@ -35,6 +38,22 @@ def test_compute_advantages(self):
 
         assert np.array_equal(adv, desired_val)
 
+    def test_normalize_pixel_observation(self):
```
you will also need to test that CategoricalCNNPolicy is doing the right think here...

maliesa96 commented:
Updated to use akro v0.0.7. Here are the major changes:

- ` normalize_pixel_observation` and `normalize_pixel_batch` no longer check the observation type. This is done by the callers by checking to see if the observation space is an `akro.Image` instance.

- Added a test to make sure `CategoricalCNNPolicy` isn't trying to normalize observations that aren't images.

- added an `is_image` flag to `GarageEnv`and `TfEnv` that gets passed to `akro.from_gym`.

ryanjulian commented:
@maliesa96 you appear to have an unclean branch now, as a results of (probably) using `git merge` or `git pull`, which we don't use in this repo.

You can solve this by using `git rebase -i master` and then squashing all but your head commit. In the future, don't use `git merge`, `git pull` or the "Update branch" button on GitHub. All of those generate merge commits, which we don't allow on master.

For more info and links to guides on how to use our git worfklow, check out the Git section of CONTRIBUTING.md.

ryanjulian commented:
I will review this as soon as you've updated the PR -- I'd like to review it now, but there are just too many touched files for me to figure out what to pay attention to.

maliesa96 commented:
Weird, I had already done a rebase and squashed the merge commits.  I had two commits that I just squashed into one, did that fix it?

maliesa96 commented:
Looks okay now I believe. My mistake, I should've used my master branch to track rlworkgroup/master and a feature branch for my changes.

ryanjulian commented:
Your branch still have conflicts which need resolving, but it looks much better now.

I understand that most people aren't used to working with multiple remotes...

ryanjulian commented on src/garage/envs/base.py:
```diff
@@ -41,10 +41,13 @@ class GarageEnv(gym.Wrapper):
             exist, a `gym.error` is thrown.
+        is_image (bool): True if observations contain pixel values,
+            false otherwise. Setting this to true causes pixel values
+            to be normalized.
```
doesn't it also cause the observation space to be cast to akro.Image?

ryanjulian commented on src/garage/tf/policies/categorical_cnn_policy.py:
```diff
@@ -101,9 +99,12 @@ def __init__(self,
+        state_input = tf.compat.v1.placeholder(tf.uint8,
                                                shape=(None, ) + self.obs_dim)
 
+        if isinstance(self.env_spec.observation_space, akro.Image):
```
a CategoricalCNNPolicy can only accept akro.Images, so I think you should remove these conditionals and instead add a guard in the constructor

```py
if not isinstance(env_spec.observation_space, akro.Image):
    raise ValueError('CategoricalCNNPolicy can only process akro.Image observations, but received an env_spec with observation_space of type {}'
    .format(type(env_spec.observation_space).__name__))
```

maliesa96 replied:
Does this apply to all CNN primitives? I was under the impression that CNNs can be used on non-image data as well.

ryanjulian replied:
You can apply a CNNs to arbitrary 2D data, which of course doesn't have to be an image (but usually are).

However, for the purposes of type-safety, CNNs operate the type `Image` (i.e. arbitrary 2D data). If, for instance, someone wants their CNNs to operate on sound data, they could rearrange that data into a 2D grid, which we call an image.

maliesa96 replied:
Shouldn't non-image 2D data be classified as a Box? Using your sound example, if one were to input sound data as an `akro.Image`,  `CategoricalCNNPolicy` will normalize this data by dividing by 255 thinking it contains pixel values.

maliesa96 replied:
On second thought this might not be an issue since an akro.Image can only contain values between 0 and 255, though it would mean non-image data would have to be mapped to that range as well.

ahtsan replied:
How about CNN could take either akro.Image or akro.Box, and only normalize when it's akro.image?

maliesa96 replied:
I agree @ahtsan. How does that sound @ryanjulian?

ryanjulian replied:
My preference is to enforce a 1:1 mapping between types and networks (i.e. CNNs take images). The purpose of this change was to simplify the type relationships in this code by eliminating `Box` types, since they are hard to validate, not introduce a new, complicated set of cases.

Perhaps the definition of `akro.Image` is too restrictive, and we should provision for float-valued images or int-valued images, but that adds complexity which also has to be justified by identifying a real use case, not a hypothetical one.

In practical terms, if someone wanted to use a CNN with float-valued data (for instance), converting that data to uint8 has shown to have a small-to-negligible impact on performance in NNs. If this isn't sufficient for them, there's nothing stopping them from writing their own CategoricalCNNPolicy based largely on the existing one.

As for what to actually do, I think accepting `Box` is fine, as long as you validate that it's a 2D or 3D `Box`. The 2D box is assumed to be shaped `(W, H)` and the 3D box is assumed to be shaped `(W, H, C)`.

ryanjulian replied:
@krzentner i'm curious to hear your thoughts

krzentner replied:
I would also lean towards accepting `Box`, since many people will have environments that don't use the `Image` type, but they would expect garage to handle.

If they chose to use a CNN, their intent is fairly clear anyways.

ryanjulian replied:
Alright -- so we can merge this when all the CNN types validate that the input is either a 2D/3D Box or an Image.

maliesa96 replied:
@ryanjulian sounds good. I assume they should also normalize images?

ryanjulian commented:
Your branch still have conflicts which need resolving, but it looks much better now.

I understand that most people aren't used to working with multiple remotes...

ryanjulian commented:
Note that you should be adding these checks to other CNN-based primitives, in both torch and tf:

tf/GaussianCNNBaseline
tf/DiscreteCNNQFunction
tf/GaussianCNNRegressor

There may be others. Any CNN primitives which accepts an EnvSpec should be checking this.

ryanjulian commented on src/garage/tf/algos/dqn.py:
```diff
@@ -240,9 +241,11 @@ def optimize_policy(self, itr, samples_data):
-        next_observations = normalize_pixel_batch(self.env_spec,
-                                                  next_observations)
+
+        if isinstance(self.env_spec.observation_space, akro.Image):
```
DQN can only use Image observations, so your code should actually just be raising an exception when you get a non-Image observation space.

maliesa96 replied:
About this, this would actually break `examples/dqn_cartpole`. The observations in that example are Boxes relating to the position of the cart and pole, not RGB pixels.

maliesa96 commented on src/garage/sampler/off_policy_vectorized_sampler.py:
```diff
@@ -93,15 +95,19 @@ def obtain_samples(self, itr, batch_size=None, whole_paths=True):
         while n_samples < batch_size:
             policy.reset(completes)
-            input_obses = self.algo.env_spec.observation_space.flatten_n(obses)
-            obs_normalized = tensor_utils.normalize_pixel_batch(
-                self._env_spec, input_obses)
+            obs_space = self.algo.env_spec.observation_space
+
+            if not isinstance(obs_space, akro.Image):
+                input_obses = obs_space.flatten_n(obses)
+            else:
+                input_obses = obses
+
```
@krzentner is this the expected behavior? Just want to make sure this change doesn't break anything.

krzentner replied:
I think it's better to fix this in the policy. (i.e. assume that inputs into the policy *might* be flattened already).

maliesa96 replied:
Fixed. The sampler now flattens all inputs, and CNN policies unflatten them.

maliesa96 commented:
Updated. Summary of changes:

- GaussianCNNBaseline, CategoricalCNNPolicy, and DiscreteCNNQFunction all only accept `akro.Image` or 2D/3D `akro.Box` spaces.

- These primitives all normalize `akro.Image` observations internally, meaning OffPolicyVectorizedSampler and DQN no longer normalize pixel values themselves.

ryanjulian commented:
@maliesa96 let us know when this is ready

maliesa96 commented on src/garage/tf/algos/dqn.py:
```diff
@@ -236,10 +242,12 @@ def optimize_policy(self, itr, samples_data):
-                                                  next_observations)
+        # optimization. DiscreteCNNQFunction already does this internally.
+
+        if not isinstance(self.qf, DiscreteCNNQFunction):
+            observations = normalize_pixel_batch(observations)
+            next_observations = normalize_pixel_batch(next_observations)
+
         loss, _ = self._train_qf(observations, actions, rewards, dones,
                                  next_observations)
```
@ryanjulian does this part look okay? the idea is to avoid normalizing twice and only normalize if the Q function doesn't do so internally (which at the moment is only done by DiscreteCNNQFunction). Alternatively I can make every Q function normalize akro.Images, though I'm not sure that's always desirable. 

If this looks good then it's ready to be merged.

ryanjulian replied:
this is hacky -- i would prefer if you make it consistent across the codebase. 

an argument can be made either way about whether the QFunction or the algorithm should be normalizing.

without doing a grand review of API design, we should just go with the decision which creates the last amount of change. in this case, that seems that it would be to remove normalization from DiscreteCNNQFunction.

ryanjulian replied:
note that #734 actually tells you what to do here -- the models (FooCNNBar) should be doing the normalizing, not the algorithms.

ryanjulian commented:
note that if your CI/GitHub checks are not passing, I'm not going to do legwork for you to finish it up and merge.

please let me know when this is ready-to-merge (i.e. everything is green)

maliesa96 commented:
Yep I know. The builds were failing due to code that isn't mine (missing docstrings in files I modified), I wasn't sure if that meant I'd have to make those fixes. I'll make the changes and push.

ryanjulian commented:
Our policy is that if you a touch a file which needs updated due to stronger standards (e.g. docstrings, style), it's your responsibility to update it.

We usually only relax this when a PR touches a ton of files all at once.

ryanjulian commented on src/garage/tf/baselines/gaussian_cnn_baseline.py:
```diff
@@ -27,6 +28,15 @@ def __init__(
             regressor_args=None,
             name='GaussianCNNBaseline',
     ):
+        assert isinstance(env_spec.observation_space, akro.Box) and \
```
an assert is not the right way to go here. that will crash the program. asserts are only for testing conditions which absolutely should not be logically possible unless something terrible has happened, not for validating user input.

Here you just want a simple `if` and raise a ValueError

ryanjulian commented on src/garage/tf/baselines/gaussian_cnn_baseline.py:
```diff
@@ -27,6 +28,15 @@ def __init__(
             regressor_args=None,
             name='GaussianCNNBaseline',
     ):
+        assert isinstance(env_spec.observation_space, akro.Box) and \
```
wait -- what if I get an `akro.Image`? is that not allowed?

maliesa96 replied:
akro.Image is a subclass of akro.Box, so that'll work fine.

ryanjulian replied:
ah, thanks for pointing that out.

ryanjulian commented on src/garage/tf/misc/tensor_utils.py:
```diff
@@ -59,8 +60,22 @@ def flatten_batch_dict(d, name=None):
+                                name=name)[1]
+
+
+def normalize_pixel_observation(obs):
```
is this really a necessary helper function? its implementation is 1-line.

ryanjulian commented on src/garage/tf/q_functions/discrete_cnn_q_function.py:
```diff
@@ -86,6 +88,15 @@ def __init__(self,
                  output_b_init=tf.zeros_initializer(),
                  dueling=False,
                  layer_normalization=False):
+        assert isinstance(env_spec.observation_space, akro.Box) and \
```
use a ValueError

ryanjulian commented on src/garage/tf/policies/categorical_cnn_policy.py:
```diff
@@ -76,6 +74,16 @@ def __init__(self,
             'CategoricalCNNPolicy only works with akro.Discrete action '
             'space.')
+
+        assert isinstance(env_spec.observation_space, akro.Box) and \
```
raise a ValueError

ryanjulian replied:
err TypeError, i mean

maliesa96 commented:
Note that the Travis builds won't pass without merging #1313.

ryanjulian commented on src/garage/misc/tensor_utils.py:
```diff
@@ -275,23 +274,19 @@ def truncate_tensor_dict(tensor_dict, truncated_len):
 
 
-def normalize_pixel_batch(env_spec, observations):
+def normalize_pixel_batch(observations):
```
@maliesa96 is this resolved?

maliesa96 replied:
Yup.

maliesa96 commented:
@ryanjulian Travis builds pass but won't run the mujoco tests because external PR containers don't include a mujoco key (explains the low code coverage). Does codecov not report the patch and project statuses when the PR is below the codecov threshold?

ryanjulian commented:
@Mergifyio rebase

ryanjulian commented:
@maliesa96 can you rebase this branch and ping me when the CI goes green? then I can merge it.

ryanjulian commented:
@maliesa96 never mind, i think i can do it myself

ryanjulian commented:
@Mergifyio rebase

ryanjulian commented:
@Mergifyio rebase
